sklearnex/cluster/k_means.py

```diff
@@ -1,17 +1,370 @@
+                        "Sample weights are not ones.",
+                    ),
+                    (
+                        is_data_supported,
+                        "Supported data formats: Dense, CSR (oneDAL version >= 2024.6.0).",
+                    ),
+                ]
+            )
+
+            return patching_status
+
+        def fit(self, X, y=None, sample_weight=None):
+            if sklearn_check_version("1.0"):
+                self._check_feature_names(X, reset=True)
+            if sklearn_check_version("1.2"):
+                self._validate_params()
+
+            dispatch(
+                self,
+                "fit",
+                {
+                    "onedal": self.__class__._onedal_fit,
+                    "sklearn": sklearn_KMeans.fit,
+                },
+                X,
+                y,
+                sample_weight,
+            )
+
+            return self
+
+        def _onedal_fit(self, X, _, sample_weight, queue=None):
+            X = self._validate_data(
+                X,
+                accept_sparse="csr",
+                dtype=[np.float64, np.float32],
+            )
+
+            if sklearn_check_version("1.2"):
+                self._check_params_vs_input(X)
+            else:
+                self._check_params(X)
+
+            self._n_features_out = self.n_clusters
+            self._n_threads = _openmp_effective_n_threads()
+
+            self._initialize_onedal_estimator()
+            self._onedal_estimator.fit(X, queue=queue)
+
+            self._save_attributes()
+
+        def _onedal_predict_supported(self, method_name, X, sample_weight=None):
+            assert method_name == "predict"
+
+            class_name = self.__class__.__name__
+            is_data_supported = (
+                _is_csr(X) and daal_check_version((2024, "P", 600))
+            ) or not issparse(X)
+            patching_status = PatchingConditionsChain(
+                f"sklearn.cluster.{class_name}.predict"
+            )
+
+            supported_algs = ["auto", "full", "lloyd", "elkan"]
+
+            patching_status.and_conditions(
+                [
+                    (
+                        self.algorithm in supported_algs,
+                        "Only lloyd algorithm is supported, elkan is computed using lloyd.",
+                    ),
+                    (
+                        is_data_supported,
+                        "Supported data formats: Dense, CSR (oneDAL version >= 2024.6.0).",
+                    ),
+                    (
+                        hasattr(self, "_onedal_estimator"),
+                        "oneDAL model was not fit.",
+                    ),
+                ]
+            )
+
+            return patching_status
+
+        if sklearn_check_version("1.5"):
+
+            @wrap_output_data
+            def predict(self, X):
+                self._check_feature_names(X, reset=False)
+                self._validate_params()
```

Comment:
_check_feature_names is a part of validate_data, which is run in sklearn fallback and in _onedal_predict and can be removed.  _validate_params should be moved to _onedal_predict. This should be done for both versions of predict.

Reply:
I think it depends on sklearn version. Make sense to add sklearn versioning here. @md-shafiul-alam please take also into account discussion here #1906